Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [28]:
import pandas as pd

In [3]:
import numpy as np

In [4]:
horse_data=pd.read_csv('horse_data.csv')
horse_data.head()

surgery  age  hospital_number rectal_temperature pulse respiratory_rate  \
0       2    1           530101              38.50    66               28   
1       1    1           534817               39.2    88               20   
2       2    1           530334              38.30    40               24   
3       1    9          5290409              39.10   164               84   
4       2    1           530255              37.30   104               35   

  temperature_of_extremities peripheral_pulse mucous_membranes  \
0                          3                3                ?   
1                          ?                ?                4   
2                          1                1                3   
3                          4                1                6   
4                          ?                ?                6   

  capillary_refill_time  ... packed_cell_volume total_protein  \
0                     2  ...              45.00          8.40   
1                     1  ...                 50            85   
2                     1  ...              33.00          6.70   
3                     2  ...              48.00          7.20   
4                     2  ...              74.00          7.40   

  abdominocentesis_appearance abdomcentesis_total_protein outcome  \
0                           ?                           ?       2   
1                           2                           2       3   
2                           ?                           ?       1   
3                           3                        5.30       2   
4                           ?                           ?       2   

  surgical_lesion type_of_lesion1 type_of_lesion2 type_of_lesion3 cp_data  
0               2           11300               0               0       2  
1               2            2208               0               0       2  
2               2               0               0               0       1  
3               1            2208               0               0       1  
4               2            4300               0               0       2  

[5 rows x 28 columns]

In [25]:
horse_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   surgery                      300 non-null    object
 1   age                          300 non-null    int64 
 2   hospital_number              300 non-null    int64 
 3   rectal_temperature           300 non-null    object
 4   pulse                        300 non-null    object
 5   respiratory_rate             300 non-null    object
 6   temperature_of_extremities   300 non-null    object
 7   peripheral_pulse             300 non-null    object
 8   mucous_membranes             300 non-null    object
 9   capillary_refill_time        300 non-null    object
 10  pain                         300 non-null    object
 11  peristalsis                  300 non-null    object
 12  abdominal_distension         300 non-null    object
 13  nasogastric_tube             300 no

In [5]:
horse_analysis = horse_data[['surgery', 'age', 'rectal_temperature', 'pulse', 'temperature_of_extremities', 'abdomen', 'abdominal_distension', 'outcome']]
horse_analysis.head()

surgery  age rectal_temperature pulse temperature_of_extremities abdomen  \
0       2    1              38.50    66                          3       5   
1       1    1               39.2    88                          ?       2   
2       2    1              38.30    40                          1       1   
3       1    9              39.10   164                          4       ?   
4       2    1              37.30   104                          ?       ?   

  abdominal_distension outcome  
0                    4       2  
1                    2       3  
2                    1       1  
3                    4       2  
4                    ?       2

числовыми величинами являются температура и пульс, все остальные (среди выбранных мною) - категорийные

In [6]:
import re

In [7]:
horse = horse_analysis.replace('\?', np.nan, regex=True)

In [8]:
horse['rectal_temperature'] = horse['rectal_temperature'].apply(pd.to_numeric, downcast='float', errors='coerce')
horse['pulse'] = horse['pulse'].apply(pd.to_numeric, downcast='float', errors='coerce')

In [9]:
horse.describe()

age  rectal_temperature       pulse
count  300.000000          240.000000  276.000000
mean     1.640000           38.167917   71.913043
std      2.173972            0.732289   28.630557
min      1.000000           35.400002   30.000000
25%      1.000000           37.799999   48.000000
50%      1.000000           38.200001   64.000000
75%      1.000000           38.500000   88.000000
max      9.000000           40.799999  184.000000

In [10]:
horse_a = horse[horse.age.between(1, 2, inclusive=True)].sort_values('rectal_temperature')
horse_a

surgery  age  rectal_temperature  pulse temperature_of_extremities  \
44        1    1           35.400002  140.0                          3   
141       2    1           36.000000   42.0                        NaN   
238       2    1           36.099998   88.0                          3   
80        1    1           36.400002   98.0                          3   
298       1    1           36.500000  100.0                          3   
..      ...  ...                 ...    ...                        ...   
274       1    1                 NaN   76.0                        NaN   
282       1    1                 NaN   70.0                          3   
288       1    1                 NaN    NaN                        NaN   
293       1    1                 NaN   78.0                          3   
295       1    1                 NaN  120.0                          4   

    abdomen abdominal_distension outcome  
44        5                  NaN       3  
141     NaN                  NaN       2  
238       4                    2       3  
80        4                    2       2  
298       4                    3       1  
..      ...                  ...     ...  
274       5                    4       3  
282       5                    2       2  
288     NaN                  NaN       1  
293       4                  NaN       3  
295       5                  NaN       3  

[276 rows x 8 columns]

In [47]:
horse_a.describe()

age  rectal_temperature       pulse
count  276.0          220.000000  254.000000
mean     1.0           38.124545   67.456693
std      0.0            0.735988   23.933953
min      1.0           35.400002   30.000000
25%      1.0           37.700001   48.000000
50%      1.0           38.099998   60.000000
75%      1.0           38.500000   84.000000
max      1.0           40.799999  140.000000

In [12]:
import matplotlib.pyplot as plt
from scipy import stats
import seaborn as sns
import statistics
import traceback

Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

числовые столбцы здесь только rectal_temperature и pulse

In [13]:
q1 = horse_a.rectal_temperature.quantile(0.25)
q3 = horse_a.rectal_temperature.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
temp = horse_a[horse_a.rectal_temperature.between(lower_bound, upper_bound, inclusive=True)]
temp

surgery  age  rectal_temperature  pulse temperature_of_extremities  \
251       2    1           36.599998   42.0                          3   
113       1    1           36.799999   60.0                        NaN   
277       2    1           36.900002   50.0                          2   
70        2    1           37.000000   60.0                          3   
292       1    1           37.000000   66.0                          1   
..      ...  ...                 ...    ...                        ...   
120       1    1           39.400002   54.0                          1   
162       2    1           39.500000   60.0                          3   
117       2    1           39.500000    NaN                          3   
231       2    1           39.500000   92.0                          3   
99        2    1           39.599998  108.0                          3   

    abdomen abdominal_distension outcome  
251       5                    1       2  
113     NaN                  NaN       2  
277     NaN                    2       1  
70        4                    3       3  
292       5                    3       2  
..      ...                  ...     ...  
120       4                    2       1  
162     NaN                    2       1  
117       5                    3       3  
231     NaN                    1       2  
99        5                    3       1  

[209 rows x 8 columns]

In [14]:
#если просто удалить, то уберется слишком большой массив данных
horse_a.sort_values('rectal_temperature')

surgery  age  rectal_temperature  pulse temperature_of_extremities  \
44        1    1           35.400002  140.0                          3   
141       2    1           36.000000   42.0                        NaN   
238       2    1           36.099998   88.0                          3   
80        1    1           36.400002   98.0                          3   
298       1    1           36.500000  100.0                          3   
..      ...  ...                 ...    ...                        ...   
274       1    1                 NaN   76.0                        NaN   
282       1    1                 NaN   70.0                          3   
288       1    1                 NaN    NaN                        NaN   
293       1    1                 NaN   78.0                          3   
295       1    1                 NaN  120.0                          4   

    abdomen abdominal_distension outcome  
44        5                  NaN       3  
141     NaN                  NaN       2  
238       4                    2       3  
80        4                    2       2  
298       4                    3       1  
..      ...                  ...     ...  
274       5                    4       3  
282       5                    2       2  
288     NaN                  NaN       1  
293       4                  NaN       3  
295       5                  NaN       3  

[276 rows x 8 columns]

In [18]:
lower_bound = q1 - (1.5 * iqr) 
lower_bound  #есть значения, которые чуть меньше, но  погуглив, я выяснила, что такая температцура у лошади нормальная

36.50000190734863

если смотреть на те строки, для которых температура ниже первого квартиля, то по столбцу 'temperature_of_extremities', значение которого равно 3 - холодные конечности, можно предположить, что выброса как такого нет

In [19]:
upper_bound = q3 + (1.5 * iqr)
upper_bound  #нет значений, которые были бы выше данной границы

39.69999885559082

In [20]:
#подведя итог по температуре, оставляем все значения
horse_a.sort_values('rectal_temperature')

surgery  age  rectal_temperature  pulse temperature_of_extremities  \
44        1    1           35.400002  140.0                          3   
141       2    1           36.000000   42.0                        NaN   
238       2    1           36.099998   88.0                          3   
80        1    1           36.400002   98.0                          3   
298       1    1           36.500000  100.0                          3   
..      ...  ...                 ...    ...                        ...   
274       1    1                 NaN   76.0                        NaN   
282       1    1                 NaN   70.0                          3   
288       1    1                 NaN    NaN                        NaN   
293       1    1                 NaN   78.0                          3   
295       1    1                 NaN  120.0                          4   

    abdomen abdominal_distension outcome  
44        5                  NaN       3  
141     NaN                  NaN       2  
238       4                    2       3  
80        4                    2       2  
298       4                    3       1  
..      ...                  ...     ...  
274       5                    4       3  
282       5                    2       2  
288     NaN                  NaN       1  
293       4                  NaN       3  
295       5                  NaN       3  

[276 rows x 8 columns]

In [22]:
q1 = horse_a.pulse.quantile(0.25)
q3 = horse_a.pulse.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
puls = horse_a[horse_a.pulse.between(lower_bound, upper_bound, inclusive=True)]
puls.sort_values('pulse')

surgery  age  rectal_temperature  pulse temperature_of_extremities  \
232       1    1           38.500000   30.0                        NaN   
130       1    1           37.799999   30.0                        NaN   
242       1    1                 NaN   36.0                          1   
203       2    1           37.200001   36.0                          1   
276       1    1           38.000000   36.0                          3   
..      ...  ...                 ...    ...                        ...   
43        1    1                 NaN  120.0                          3   
295       1    1                 NaN  120.0                          4   
172       1    1           38.500000  129.0                          3   
63        2    1           38.200001  130.0                          4   
148       1    1           38.299999  132.0                        NaN   

    abdomen abdominal_distension outcome  
232     NaN                  NaN       1  
130     NaN                  NaN       2  
242       4                    3       2  
203       1                    1       1  
276     NaN                    2       3  
..      ...                  ...     ...  
43        5                    4       3  
295       5                  NaN       3  
172     NaN                    3       1  
63      NaN                    4       3  
148       4                    2       1  

[253 rows x 8 columns]

In [23]:
lower_bound = q1 - (1.5 * iqr) 
lower_bound  #нет значений, которые были бы ниже данной границы

-6.0

In [24]:
upper_bound = q3 + (1.5 * iqr)
upper_bound

138.0

повышенный пульс (max которого реально может достигать 250) должен говорить о том, что у лошади либо шок, либо инфекция)
для строки с индексом 44 (пульс = 140) значение столбца 'abdomen' =5, что означает раздутую толстую кишку, т.е. инфекция какая-то, поэтому пульс действительно может быть повышенным

In [26]:
horse_a.sort_values('pulse', ascending=False)

surgery  age  rectal_temperature  pulse temperature_of_extremities  \
44        1    1           35.400002  140.0                          3   
148       1    1           38.299999  132.0                        NaN   
63        2    1           38.200001  130.0                          4   
172       1    1           38.500000  129.0                          3   
273       1    1           37.700001  120.0                          3   
..      ...  ...                 ...    ...                        ...   
174       2    1                 NaN    NaN                          3   
175       1    1                 NaN    NaN                        NaN   
207       1    1                 NaN    NaN                        NaN   
216       2    1                 NaN    NaN                        NaN   
288       1    1                 NaN    NaN                        NaN   

    abdomen abdominal_distension outcome  
44        5                  NaN       3  
148       4                    2       1  
63      NaN                    4       3  
172     NaN                    3       1  
273     NaN                    3       2  
..      ...                  ...     ...  
174     NaN                    4       2  
175     NaN                  NaN       1  
207     NaN                  NaN       2  
216     NaN                  NaN       1  
288     NaN                  NaN       1  

[276 rows x 8 columns]

Таким образом, выбросов здесь нет.

Задание 3. Работа с пропусками
Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [29]:
for col in horse_a.columns:
    pct_missing = horse_a[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery - 0.4%
age - 0.0%
rectal_temperature - 20.3%
pulse - 8.0%
temperature_of_extremities - 17.4%
abdomen - 35.5%
abdominal_distension - 17.0%
outcome - 0.4%


поскольку очень мало пропусков в 1ом и последнем столбце, то посмотрим на них

In [30]:
horse_a.surgery.value_counts()

1    162
2    113
Name: surgery, dtype: int64

In [32]:
horse_a.outcome.value_counts()

1    166
2     66
3     43
Name: outcome, dtype: int64

In [31]:
horse_a.sort_values('surgery', ascending=False)

surgery  age  rectal_temperature  pulse temperature_of_extremities  \
89        2    1           38.299999   42.0                          1   
119       2    1           38.099998   56.0                          2   
90        2    1           38.000000   52.0                        NaN   
91        2    1           40.299999  114.0                          3   
54        2    1           40.299999  114.0                          3   
..      ...  ...                 ...    ...                        ...   
268       1    1           38.200001   42.0                        NaN   
100       1    1           38.200001   40.0                          3   
146       1    1           38.200001   48.0                          1   
295       1    1                 NaN  120.0                          4   
132     NaN    1           38.000000   48.0                          3   

    abdomen abdominal_distension outcome  
89      NaN                    1       1  
119     NaN                    1       1  
90        1                  NaN       1  
91        5                    3       2  
54        5                    3       3  
..      ...                  ...     ...  
268     NaN                  NaN       1  
100       1                  NaN       1  
146       5                    4       1  
295       5                  NaN       3  
132       5                    1     NaN  

[276 rows x 8 columns]

 отсутствуют значения для строки 132. У лошади температура и пульс в норме, но 'abdomen'=5 - раздутая толстая кишка, значит, ей должны были делать операцию, поскольку 166 лошадей выжили, то пусть и эта тоже выживит))) (просто назначим ей моду)

In [38]:
horse_a.surgery.fillna(horse_a.surgery.mode()[0], inplace=True)

surgery  age  rectal_temperature  pulse temperature_of_extremities  \
89        2    1           38.299999   42.0                          1   
119       2    1           38.099998   56.0                          2   
91        2    1           40.299999  114.0                          3   
54        2    1           40.299999  114.0                          3   
281       2    1           40.000000   78.0                          3   
..      ...  ...                 ...    ...                        ...   
283       1    1           38.200001   72.0                        NaN   
192       1    1           38.200001   48.0                          1   
268       1    1           38.200001   42.0                        NaN   
100       1    1           38.200001   40.0                          3   
295       1    1                 NaN  120.0                          4   

    abdomen abdominal_distension outcome  
89      NaN                    1       1  
119     NaN                    1       1  
91        5                    3       2  
54        5                    3       3  
281       1                    1       2  
..      ...                  ...     ...  
283     NaN                  NaN       1  
192     NaN                    3       1  
268     NaN                  NaN       1  
100       1                  NaN       1  
295       5                  NaN       3  

[276 rows x 8 columns]

In [39]:
horse_a.outcome.fillna(horse_a.outcome.mode()[0], inplace=True)

In [41]:
#поскольку у лошадей температура почти у всех в норме, то заполним модой
horse_a.rectal_temperature.fillna(horse_a.rectal_temperature.mode()[0], inplace=True)

In [42]:
horse_a.abdomen.value_counts()

5    78
4    42
1    26
2    19
3    13
Name: abdomen, dtype: int64

In [43]:
horse_a.abdominal_distension.value_counts()

1    73
2    62
3    60
4    34
Name: abdominal_distension, dtype: int64

In [51]:
#поскольку строгой взаимосвязи нет для столбцов abdomen, abdominal_distension и temperature_of_extremities, то заполним модой
horse_a.abdominal_distension.fillna(horse_a.abdominal_distension.mode()[0], inplace=True)
horse_a.abdomen.fillna(horse_a.abdomen.mode()[0], inplace=True)
horse_a.temperature_of_extremities.fillna(horse_a.temperature_of_extremities.mode()[0], inplace=True)

In [53]:
print(horse_a.groupby('rectal_temperature').pulse.median())

rectal_temperature
35.400002    140.0
36.000000     42.0
36.099998     88.0
36.400002     98.0
36.500000     89.0
36.599998     42.0
36.799999     60.0
36.900002     50.0
37.000000     63.0
37.099998     75.0
37.200001     60.0
37.299999     49.0
37.400002     84.0
37.500000     48.0
37.599998     64.0
37.700001     66.0
37.799999     52.0
37.900002     60.0
38.000000     65.5
38.099998     58.0
38.200001     56.0
38.299999     53.0
38.400002     54.0
38.500000     76.0
38.599998     48.0
38.700001     56.0
38.799999     50.0
38.900002     80.0
39.000000     64.0
39.099998     72.0
39.200001     88.0
39.299999     64.0
39.400002    110.0
39.500000     76.0
39.599998    108.0
39.900002     72.0
40.000000     78.0
40.299999    114.0
40.799999     72.0
Name: pulse, dtype: float64


In [54]:
#поскольку и здесь нет корреляции между температурой и пульсом, то
horse_a.pulse.fillna(horse_a.pulse.mode()[0], inplace=True)

In [55]:
horse_a

surgery  age  rectal_temperature  pulse temperature_of_extremities  \
44        1    1           35.400002  140.0                          3   
141       2    1           36.000000   42.0                          3   
238       2    1           36.099998   88.0                          3   
80        1    1           36.400002   98.0                          3   
298       1    1           36.500000  100.0                          3   
..      ...  ...                 ...    ...                        ...   
274       1    1           38.000000   76.0                          3   
282       1    1           38.000000   70.0                          3   
288       1    1           38.000000   48.0                          3   
293       1    1           38.000000   78.0                          3   
295       1    1           38.000000  120.0                          4   

    abdomen abdominal_distension outcome  
44        5                    1       3  
141       5                    1       2  
238       4                    2       3  
80        4                    2       2  
298       4                    3       1  
..      ...                  ...     ...  
274       5                    4       3  
282       5                    2       2  
288       5                    1       1  
293       4                    1       3  
295       5                    1       3  

[276 rows x 8 columns]

In [56]:
horse_a.describe()

age  rectal_temperature       pulse
count  276.0          276.000000  276.000000
mean     1.0           38.099275   65.905797
std      0.0            0.658704   23.555856
min      1.0           35.400002   30.000000
25%      1.0           37.799999   48.000000
50%      1.0           38.000000   60.000000
75%      1.0           38.400002   80.500000
max      1.0           40.799999  140.000000

In [57]:
for col in horse_a.columns:
    pct_missing = horse_a[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery - 0.0%
age - 0.0%
rectal_temperature - 0.0%
pulse - 0.0%
temperature_of_extremities - 0.0%
abdomen - 0.0%
abdominal_distension - 0.0%
outcome - 0.0%
